packages/near-lake-primitives/test/block.test.ts

```diff
@@ -1,22 +1,80 @@
-import { readFile } from 'fs/promises';
+import { readFile } from "fs/promises";
 
-import { Block } from '../src/types';
+import { Block } from "../src/types";
 
-describe('Block', () => {
-  it('serializes meta transactions', async () => {
-    let streamerMessageBuffer = await readFile(`${__dirname}/../../../blocks/105793821.json`);
+import { fromBorsh } from "../src/fromBorsh";
+
+describe("Block", () => {
+  it("serializes meta transactions", async () => {
+    let streamerMessageBuffer = await readFile(
+      `${__dirname}/../../../blocks/105793821.json`
+    );
     let streamerMessage = JSON.parse(streamerMessageBuffer.toString());
     let block = Block.fromStreamerMessage(streamerMessage);
 
-    const actions = block.actionByReceiptId('Dpego7SpsK36PyXjUMrFoSze8ZpNsB9xhb3XJJYtXSix');
+    const actions = block.actionByReceiptId(
+      "Dpego7SpsK36PyXjUMrFoSze8ZpNsB9xhb3XJJYtXSix"
+    );
     expect(actions?.operations[0]).toMatchSnapshot();
   });
 
-  it('parses event logs', async () => {
-    let streamerMessageBuffer = await readFile(`${__dirname}/../../../blocks/61321189.json`);
+  it("parses event logs", async () => {
+    let streamerMessageBuffer = await readFile(
+      `${__dirname}/../../../blocks/61321189.json`
+    );
     let streamerMessage = JSON.parse(streamerMessageBuffer.toString());
     let block = Block.fromStreamerMessage(streamerMessage);
 
     expect(block.events()).toMatchSnapshot();
-  })
+  });
+
+  function base64toHex(encodedValue: string) {
+    let buff = Buffer.from(encodedValue, "base64");
+    return buff.toString("hex");
+  }
+
+  it("deserializes using borsch", async () => {
+    let streamerMessageBuffer = await readFile(
+      `${__dirname}/../../../blocks/114158749.json`
+    );
+    let streamerMessage = JSON.parse(streamerMessageBuffer.toString());
+    let block = Block.fromStreamerMessage(streamerMessage);
+
+    const stateChanges = block.streamerMessage.shards
+      .flatMap((e) => e.stateChanges)
+      .filter(
+        (stateChange) =>
+          stateChange.change.accountId === "devgovgigs.near" &&
+          stateChange.type === "data_update"
+      );
+
+    const addOrEditPost = stateChanges
+      .map((stateChange) => stateChange.change)
+      .filter((change) => base64toHex(change.keyBase64).startsWith("05"))
+      .map((c) => ({
+        k: Buffer.from(c.keyBase64, "base64"),
+        v: Buffer.from(c.valueBase64, "base64"),
+      }));
+
+    const authorToPostId = Object.fromEntries(
```

Comment:
Thanks @Tguntenaar ! I actually think it would be highly beneficial to be able to use complex schema definition in the implementation, so that you can take it almost directly from your smart contract code and plug here. Do I get it right that this will essentially eliminate the need for this code that essentially implements it manually?

Reply:
Hi @pkudinov, Yes that is right! Instead of fishing the `post_id` and the `author` out of the buffer by hand we can deserialize the entire state in one time and read it as a object.

I will also add the `bosher` package to facilitate reproducing the format of the contract state more easily.

Reply:
@pkudinov As I was reconstructing the borsh schema. I found out that in this case it is actually easier to calculate it by hand because the `post_id` is in the beginning of the structure. While the post structure is quiet complicated to rebuild as show below.

However for other use cases including `borsher` is still usefull for deserialising so I included it in the primitives packages.

<details>
<summary>See here</summary>

```js
const postSnapshot = BorshSchema.Struct({
       // Reconstruction this schema is more compilcated than counting which bytes to skip if the property is in the beginning 
});

const addPostSchema = BorshSchema.Struct({
      id: BorshSchema.u64,
      author_id: BorshSchema.String,
      likes: BorshSchema.HashSet(
        BorshSchema.Struct({
          author_id: BorshSchema.String,
          timestamp: BorshSchema.u64,
        })
      ),
      snapshot: postSnapshot,
      snapshot_history: BorshSchema.Vec(postSnapshot),
});
```
</details>